<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-библиотек" data-toc-modified-id="Загрузка-библиотек-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка библиотек</a></span></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Просмотр-содержимого-таблиц" data-toc-modified-id="Просмотр-содержимого-таблиц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Просмотр содержимого таблиц</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-количество-обзоров-и-среднюю-оценку-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Для каждой книги количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг,-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг,-толще-50-страниц-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг, толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг,-в-которых-50-и-более-оценок" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг,-в-которых-50-и-более-оценок-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг, в которых 50 и более оценок</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Анализ базы данных крупного сервиса для чтения книг по подписке

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

<b>Задача </b>

Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Необходимо узнать:
- количество книг, вышедших после 1 января 2000 года
- для каждой книги количество обзоров и среднюю оценку
- издательство, которое выпустило наибольшее число книг, толще 50 страниц
- автора с самой высокой средней оценкой книг, в которых 50 и более оценок
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок


<b> Таблица books </b> - содержит данные о книгах:
- book_id — идентификатор книги
- author_id — идентификатор автора
- title — название книги
- num_pages — количество страниц
- publication_date — дата публикации книги
- publisher_id — идентификатор издателя

<b> Таблица authors </b> - содержит данные об авторах:
- uthor_id — идентификатор автора
- author — имя автора

<b> Таблица publishers </b> - содержит данные об издательствах:
- publisher_id — идентификатор издательства
- publisher — название издательства

<b> Таблица ratings </b> - содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки
- book_id — идентификатор книги
- username — имя пользователя, оставившего оценку
- rating — оценка книги

<b> Таблица reviews </b> - содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора
- book_id — идентификатор книги
- username — имя автора обзора
- text — текст обзора



## Загрузка библиотек

In [1]:
! pip install sqlalchemy

In [2]:
import pandas as pd
from sqlalchemy import text, create_engine

## Подключение к базе данных

In [3]:
db_config = {'user':'praktikum_student',
            'pwd':'Sdf4$2;d-d30pp',
            'host':'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port':6432,
            'db':'data-analyst-final-project-db'}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [4]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

## Просмотр содержимого таблиц

### Таблица books

In [5]:
query = '''SELECT * FROM books LIMIT 5'''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Таблица authors

In [6]:
query = '''SELECT * FROM authors LIMIT 5'''
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Таблица publishers

In [7]:
query = '''SELECT * FROM publishers LIMIT 5'''
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Таблица ratings

In [8]:
query = '''SELECT * FROM ratings LIMIT 5'''
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### Таблица reviews

In [9]:
query = '''SELECT * FROM reviews LIMIT 5'''
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Количество книг, вышедших после 1 января 2000 года

In [10]:
query = '''
    SELECT COUNT(book_id) 
    FROM books
    WHERE publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 вышло 819 книг

## Для каждой книги количество обзоров и среднюю оценку

In [29]:
query = '''
    SELECT b.book_id,
           b.title,
           COUNT(DISTINCT rev.review_id) AS count_rev,
           ROUND(AVG(rat.rating),2) AS avg_rating
    FROM books AS b
    LEFT JOIN reviews AS rev ON b.book_id=rev.book_id
    LEFT JOIN ratings AS rat ON b.book_id=rat.book_id
    GROUP BY b.book_id
    ORDER BY avg_rating DESC
'''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  count_rev  \
0         86      Arrows of the Queen (Heralds of Valdemar  #1)          2   
1        901  The Walking Dead  Book One (The Walking Dead #...          2   
2        390                                    Light in August          2   
3        972  Wherever You Go  There You Are: Mindfulness Me...          2   
4        136  Captivating: Unveiling the Mystery of a Woman'...          2   
..       ...                                                ...        ...   
995      915  The World Is Flat: A Brief History of the Twen...          3   
996      316                  His Excellency: George Washington          2   
997      202                                      Drowning Ruth          3   
998      371                                              Junky          2   
999      303                               Harvesting the Heart          2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 4 columns]

In [30]:
query = '''
    SELECT b.book_id,
           b.title,
           COUNT(DISTINCT rev.review_id) AS count_rev,
           ROUND(AVG(rat.rating),2) AS avg_rating
    FROM books AS b
    LEFT JOIN reviews AS rev ON b.book_id=rev.book_id
    LEFT JOIN ratings AS rat ON b.book_id=rat.book_id
    GROUP BY b.book_id
    ORDER BY count_rev DESC
'''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  count_rev  \
0        948                            Twilight (Twilight  #1)          7   
1        963                                Water for Elephants          6   
2        734                                   The Glass Castle          6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...          6   
4        695  The Curious Incident of the Dog in the Night-Time          6   
..       ...                                                ...        ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel          0   
996      808                            The Natural Way to Draw          0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites          0   
998      221                          Essential Tales and Poems          0   
999      191  Disney's Beauty and the Beast (A Little Golden...          0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

Максимальный рейтинг у книг - 5, минимальный - 1,5. 

Максимальное количество обзоров - 7, и также присутствуют книги без обзоров.

## Издательство, которое выпустило наибольшее число книг, толще 50 страниц

In [12]:
query = '''
    SELECT p.publisher_id,
           p.publisher,
           COUNT(DISTINCT b.book_id)  AS count_book
    FROM books AS b
    INNER JOIN publishers AS p ON b.publisher_id=p.publisher_id
    WHERE b.num_pages > 50
    GROUP BY p.publisher_id
    ORDER BY count_book DESC
    LIMIT 1
'''
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  count_book
0           212  Penguin Books          42

Наибольшее количество книг (42), толще 50 страниц, у издательства Penguin Books

## Автор с самой высокой средней оценкой книг, в которых 50 и более оценок

In [13]:
query = '''
    SELECT a.author,
           ROUND(AVG(r.rating),2) AS avg_rating
    FROM ratings AS r
    INNER JOIN books AS b ON r.book_id=b.book_id
    INNER JOIN authors AS a ON b.author_id=a.author_id
    WHERE r.book_id IN (SELECT r.book_id
                        FROM ratings AS r
                        GROUP BY r.book_id
                        HAVING COUNT(r.rating_id)>=50
                        )
    GROUP BY a.author_id
    ORDER BY avg_rating DESC
    LIMIT 1
    

'''
pd.io.sql.read_sql(sql=text(query), con = con)


author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автор с самой высокой средней оценкой книг, в которых больше или равно 50 оценок - J.K. Rowling/Mary GrandPré

## Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [14]:
query = '''
    SELECT ROUND(AVG(count_review),2) AS avg_count_review
    FROM (
          SELECT rev.username,
                 COUNT(rev.review_id) AS count_review
          FROM reviews AS rev
          GROUP BY rev.username
          HAVING rev.username IN (
                                  SELECT rat.username
                                  FROM ratings AS rat
                                  GROUP BY rat.username
                                  HAVING COUNT(rat.rating_id)>48
                                  ) 
          ) AS count_review
'''
pd.io.sql.read_sql(sql=text(query), con = con)

avg_count_review
0              24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, равно 24.

## Вывод

По результатам исследования мы получили, что:
- после 1 января 2000 вышло 819 книг.
- всего в каталоге 1000 книг. Максимальный рейтинг - 5, минимальный - 1,5. Количество обзоров от 0 до 7.
- издательство Penguin Books выпустило  наибольшее количество книг (42 шт.), толще 50 страниц.
- наиболее популярный автор - это союз Джоан Роулинг, британская писательница, написавшая "Гарри Поттера" и американский иллюстратор Мари Грандпре, которая делала иллюстрации к обложкам и главам книг.
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок, равно 24. Это значит они много читают и много делятся информацией о книгах.